Use require for js and replace mocks with real file tests

chrisblossom opened the pull request:
This PR replaces ``require-from-string`` with ``require`` and replaces filesystem mocks with real file tests. @davidtheclark if you aren't happy with this implementation, feel free to close.

Working from this commit I have the majority of the ``loadJs`` feature added, but it looks like that will be superseded with #126. I'll hold off with that feature until #126 is decided on, unless you think otherwise?

chrisblossom commented on src/loader.js:
```diff
@@ -12,7 +12,20 @@ function fileContentToResult(
+      config: undefined,
+      isEmpty: true,
+    };
+  }
```
When switching to real file tests I found that the content wouldn't return an empty string because of new lines. Pretty sure this was a bug.

chrisblossom replied:
``trim`` worked perfectly. I [added a test](https://github.com/davidtheclark/cosmiconfig/pull/127/files#diff-7a54a2eed8c64ebde960aa063ef05ba3R476) to confirm.

sudo-suhas commented on test/util.js:
```diff
@@ -1,55 +1,86 @@
-  };
+  constructor() {
+    const tempDir = fs.realpathSync(os.tmpdir());
+    const parent = parentModule();
```
Hey @chrisblossom, could you please explain the significance of these 2 lines?

chrisblossom replied:
I've [added some comments](https://github.com/davidtheclark/cosmiconfig/pull/127/files#diff-d8642304b1d7b6801be66d60ee3cbc5bR13). Please let me know if that clears things up.

chrisblossom replied:
The idea is to have a consistent directory name for each test for things like ``jest --watch`` and other third party tools to play nicely.

Also, because ``tmp`` doesn't use ``realpathSync``, it will break any tests [using ``process.chdir``](https://github.com/davidtheclark/cosmiconfig/pull/127/files#diff-4b7ed0b193d73a1b1fa66c79c737b673R463).

sudo-suhas replied:
I see, thanks for explaining.

sudo-suhas commented on test/util.js:
```diff
@@ -1,55 +1,86 @@
+    makeDir.sync(dirname);
+  }
+
+  createFile(file: string, contents: string | Object) {
```
Contents is always a string right? `${contents}` would not work well either if it was an object.

chrisblossom replied:
Yes, good catch! `Object` was added from a previous implementation that I forgot to remove.

sudo-suhas commented:
> This looks great! Thanks @chrisblossom.

Agreed 👍!

chrisblossom commented on test/caches.test.js:
```diff
@@ -1,456 +1,439 @@
-    util.assertSearchSequence(readFileMock, [
+  const searchPath = temp.absolutePath('a/b/c/d/e');
+  const checkResult = (readFileSpy, result) => {
+    const searchPath = temp.getSpyPathCalls(readFileSpy);
```
Do you mean only in ``caches.test.js`` or for all variables named ``searchPath`` in all tests files?

davidtheclark replied:
I meant in all the places this pattern (`getSpyPathCalls`) is used.

chrisblossom replied:
Switched them all to ``filesChecked``.

chrisblossom commented:
I've updated the PR to reflect the requested changes and fixed the tests. I reverted the updated Jest change so it will work on node 4 and [updated `getSpyPathCalls` to normalize windows directory separators](https://github.com/davidtheclark/cosmiconfig/pull/127/files#diff-d8642304b1d7b6801be66d60ee3cbc5bR63).

Also, `v5` might be a good time to drop `node 4` support. [The end of life is scheduled for the end of this month](https://github.com/nodejs/Release#release-schedule). People wanting to support `node 4` could continue to use `cosmiconfig 4` as they are currently. Maybe should be bought up as a separate issue?

Regarding this PR, what do you think about replacing all files in ``fixtures`` to use the same ``TempDir`` test pattern?

davidtheclark commented:
> Also, cosmiconfig v5 might be a good time to drop node 4 support. The end of life is scheduled for the end of this month. People wanting to support node 4 could continue to use cosmiconfig v4 as they are currently. Maybe this should be bought up as a separate issue?

Yeah, a separate, focused issue makes sense. Because cosmiconfig is low-level, I'd rather that cosmiconfig not force higher-level tools, its users, to stop supporting Node 4 (if those higher-level users have not already stopped). I know that Prettier still supports Node 4, for example. So unless absolutely necessary (and I don't think it should be), I'd prefer to keep supporting Node 4 until we're pretty confident that the bulk of cosmiconfig users have already dropped support.

> Regarding this PR, what do you think about replacing all files in fixtures to use the same TempDir test pattern?

Yeah, that's a good idea 👍

chrisblossom commented:
`fixtures` removed and the rest of the tests migrated to `TempDir`. [Found another place](https://github.com/davidtheclark/cosmiconfig/pull/127/files#diff-548f96d3487b36e2d92e138c870c529bR12) where content needed to be checked for whitespace too.

Anything else that I should do for this PR?

sudo-suhas commented on test/util.js:
```diff
@@ -1,55 +1,101 @@
+       * Replace Windows backslash directory separators with forward slashes
+       * so expected paths will be consistent cross platform
+       */
+      const normalizeCrossPlatform = relativePath.replace(/\\/g, '/');
```
Can we use [`slash`](https://github.com/sindresorhus/slash) instead?

chrisblossom replied:
I use ``slash`` in my own projects and considered adding it here, but I don't think it is worth (having to track updates etc) adding a new package for one line of code. ``slash`` has [guards against unicode and extended paths](https://github.com/sindresorhus/slash/blob/6fd50ff0bcb68812df18d526c0c3d5740fd13382/index.js#L3-L8), which isn't an issue with these tests. Do you disagree?

Also the latest version of ``slash`` doesn't support node v4 so if somehow there was an update (not sure what they could possibly change), we wouldn't benefit from it.

davidtheclark replied:
Yeah, I don't think we need it unless it solves a problem we have.

sudo-suhas commented on test/util.js:
```diff
@@ -1,55 +1,101 @@
-  startCount
-) {
-  startCount = startCount || 0;
+    (this: any).absolutePath = this.absolutePath.bind(this);
```
Would it be better to use and return a closure instead of the class? We could then avoid having to bind the methods.

chrisblossom replied:
Doesn't really matter to me. I think if a binding is required in the class, it is also required in a closure. That said, I can also just remove the bindings because they aren't required with the current implementations. I've just always considered it best practice to keep them in there to guard against anything unexpected happening.

I guess the "trade-off" with a closure would be another level of indentation or having to return the functions.

Just let me know if I should convert it to a closure, remove the bindings, or leave as is.

davidtheclark replied:
I'd also be ok with merging this and creating follow-up issues if we want to refactor. I'm hoping that next week I'll have time to get to work on #126, and I'd love to have these test changes merged in before I launch into it.

sudo-suhas replied:
Okay, let's merge this in.
